engine/access/rest/websockets/controller.go

```diff
@@ -39,62 +43,134 @@ func NewWebSocketController(
 	}
 }
 
-// HandleConnection manages the WebSocket connection, adding context and error handling.
+// HandleConnection manages the lifecycle of a WebSocket connection,
+// including setup, message processing, and graceful shutdown.
+//
+// Parameters:
+// - ctx: The context for controlling cancellation and timeouts.
 func (c *Controller) HandleConnection(ctx context.Context) {
-	//TODO: configure the connection with ping-pong and deadlines
+
+	// configuring the connection with appropriate read/write deadlines and handlers.
+	err := c.configureKeepalive()
+	if err != nil {
+		// TODO: add error handling here
+		c.logger.Error().Err(err).Msg("error configuring keepalive connection")
+		c.shutdownConnection()
+		return
+	}
+
 	//TODO: spin up a response limit tracker routine
-	go c.readMessagesFromClient(ctx)
-	c.writeMessagesToClient(ctx)
+
+	// for track all goroutines and error handling
+	g, gCtx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		return c.readMessagesFromClient(gCtx)
+	})
+
+	g.Go(func() error {
+		return c.keepalive(gCtx)
+	})
+
+	g.Go(func() error {
+		return c.writeMessagesToClient(gCtx)
+	})
+
+	if err = g.Wait(); err != nil {
+		//TODO: add error handling here
+		c.logger.Error().Err(err).Msg("error detected in one of the goroutines")
+	}
+
+	c.shutdownConnection()
```

Comment:
nit: call this with a defer at the top of the function, then remove from the error block on line 55.

Reply:
[cfad79f](https://github.com/onflow/flow-go/pull/6757/commits/cfad79fe97069d6acb12321bbd7666234c263b22)